feat!: support streaming connection and initialization fixes

sago2k8 opened the pull request:
<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR
<!-- add the description of the PR here -->

- Updates the use of launch darkly client in the browser.
- Fixes couple of bugs when changing the context and fetching the variations.
- Updates the dependecies  so the initalizaiton of the LD client. happens in the provider.
- expose events for the client to be consumed

### Notes
<!-- any additional notes for this PR -->

This is used for our team at [Redpanda](https://redpanda.com/), we use open feature across the stack and we want  to use it in our frontend apps too.

### Follow-up Tasks
<!-- anything that is related to this PR but not done here should be noted under this section -->
<!-- if there is a need for a new issue, please link it here -->
- update client references so the web-sdk can listen to events to propagate real time changes:  I might have missed it but I didn't find a way to make the web-sdk react to real time events from the provider.

### Related issues:
- #444 
### How to test
<!-- if applicable, add testing instructions under this section -->

toddbaert commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -29,78 +32,114 @@ function wrongTypeResult<T>(value: T): ResolutionDetails<T> {
   };
 }
 
+/*
+ * Provider Status
+ * defined here https://github.com/open-feature/js-sdk/blob/d26c4eeedde9cc321c69375594f1ccbfaff41a28/packages/shared/src/types.ts#L265
+ * ideally will be exported by '@openfeature/web-sdk'
+ * */
+enum ProviderStatus {
+  NOT_READY = 'NOT_READY',
+  READY = 'READY',
+  ERROR = 'ERROR',
+}
+
```
This is exported in the last few versions of the web-sdk. You can feel free to update to the latest web-sdk in the root package json of this project.

sago2k8 replied:
I updated the package version to `@openfeature/web-sdk`, and I still get a compilation error. 
```Error during bundle: Error: libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:13:3 - error TS2305: Module '"@openfeature/web-sdk"' has no exported member 'ProviderStatus'.

13   ProviderStatus,```

wondering if I am missing something, I see it in the `common` package.

toddbaert replied:
Perhaps try deleting your node modules, and trying again? Locally, when I install this, I can import `ProviderStatus` without issue. When I `go to definition` in my IDE, I see:

![image](https://github.com/open-feature/js-sdk-contrib/assets/25272906/5928b79a-38de-432a-8ea7-056122a10bcc)

Which seems right to me.

toddbaert replied:
You also might have to clear any node modules in the library folder. I generally don't keep these around and try to avoid installing them at all.

sago2k8 replied:
Indeed, thank you !!

toddbaert commented on libs/providers/launchdarkly-client/package.json:
```diff
@@ -7,7 +7,7 @@
   "peerDependencies": {
-    "@openfeature/web-sdk": "~0.3.2-experimental",
-    "launchdarkly-js-client-sdk": "^3.1.3"
+    "@openfeature/web-sdk": "*",
```
Once we release a non-experimental version, we should properly set this. I'll make an issue to do this for all components in the contribs

toddbaert commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -29,78 +32,114 @@ function wrongTypeResult<T>(value: T): ResolutionDetails<T> {
+   * implement status field/accessor
+   * https://openfeature.dev/specification/sections/providers#requirement-242
+   * */
+  public status: ProviderStatus = ProviderStatus.NOT_READY;
```
An [accessor](https://www.typescriptlang.org/docs/handbook/classes.html#accessors) might be better there. There's no need to allow modification of these field from the outside.

sago2k8 replied:
Done !!

toddbaert commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -29,78 +32,114 @@ function wrongTypeResult<T>(value: T): ResolutionDetails<T> {
+   * */
+  public status: ProviderStatus = ProviderStatus.NOT_READY;
+
+  constructor(private readonly envKey: string, { providerLogger, ...ldOptions }: LaunchDarklyProviderOptions) {
```
Though it's a breaking change, this is a much more idiomatic API. The previous one made sense before the events spec had fully matured, but now its ergonomics are not ideal

I suppose it would be possible to support the old one as well, but I question the value of doing so.

toddbaert commented on libs/providers/launchdarkly-client/src/lib/translate-result.ts:
```diff
@@ -1,5 +1,5 @@
-import {ErrorCode, ResolutionDetails} from "@openfeature/js-sdk";
-import {LDEvaluationDetail} from "launchdarkly-js-client-sdk";
+import { ErrorCode, ResolutionDetails } from '@openfeature/web-sdk';
```
nice catch here.

toddbaert commented on libs/providers/launchdarkly-client/.eslintrc.json:
```diff
@@ -1,6 +1,6 @@
 {
   "extends": ["../../../.eslintrc.json"],
-  "ignorePatterns": ["!**/*"],
+  "ignorePatterns": ["!**/*", "node_modules"],
```
Did you get lint errors in the node_modules?

If so, maybe we should move this to the root config?

sago2k8 replied:
this is the exact error I get: 
```
  Failed to load parser 'babel-eslint' declared in 'libs/providers/launchdarkly-client/node_modules/launchdarkly-js-sdk-common/.eslintrc.yaml': Cannot find module 'babel-eslint'
```
seems that eslint is trying to pick up some modules configuration

toddbaert replied:
I think running `npm install` in the library folder might be related to your other issue [here](https://github.com/open-feature/js-sdk-contrib/pull/418#discussion_r1252274079).

toddbaert commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -29,78 +32,114 @@ function wrongTypeResult<T>(value: T): ResolutionDetails<T> {
+    /*
+     * set event listeners to propagate ld events to open feature provider
+     * */
+    this.setListeners();
```
Perhaps this should be optional, based on a configuration?

I'm not sure if all users would make use of this or not, or if there's performance/traffic costs associated with setting up such listeners in LD.

toddbaert reviewed (approved):
These look like solid improvements to me. I have a few small concerns, particularly [this](https://github.com/open-feature/js-sdk-contrib/pull/418/files#r1252214568).

@sago2k8 Based on [these policies](https://github.com/open-feature/.github/blob/main/CONTRIBUTING.md#contrib-repositories), we will soon be adding a CODEOWNERs file for each sub component in these monorepos. It's clear you understand the SDK as well as this provider. Would you like to be added to the CODEOWNERS for this component?

I'm awaiting additional reviews, particularly from @kinyoklion  and @Mateoc

sago2k8 commented:
> These look like solid improvements to me. I have a few small concerns, particularly [this](https://github.com/open-feature/js-sdk-contrib/pull/418/files#r1252214568).
> 
> @sago2k8 Based on [these policies](https://github.com/open-feature/.github/blob/main/CONTRIBUTING.md#contrib-repositories), we will soon be adding a CODEOWNERs file for each sub component in these monorepos. It's clear you understand the SDK as well as this provider. Would you like to be added to the CODEOWNERS for this component?
> 
> I'm awaiting additional reviews, particularly from @kinyoklion and @Mateoc

Sure, Happy to be part of the CODEOWERS

toddbaert commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -29,78 +32,125 @@ function wrongTypeResult<T>(value: T): ResolutionDetails<T> {
+  private setListeners() {
+    this.client.on('change', (changeset: LDFlagChangeset) => {
+      this.events.emit(ProviderEvents.ConfigurationChanged, {
+        flagKeysChanged: Object.keys(changeset),
```
```suggestion
        flagsChanged: Object.keys(changeset),
```

This resolves a compilation issue for me (and the CI) which I think is related to the other issues mentioned in this PR. I _think_ an old version of the web-sdk is being cached _somewhere_ in your workspace :grimacing:

toddbaert commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-provider-options.ts:
```diff
@@ -36,5 +36,5 @@ export interface LaunchDarklyProviderOptions {
    * provide the logger in the SDK options as well.
    */
-  logger?: LDLogger  | Logger;
+  providerLogger?: LDLogger  | Logger;
```
personally, I think this is clear enough and we can just keep "logger" but that's a nit.

sago2k8 replied:
Agree!!

sago2k8 commented:
Thanks for the review @toddbaert, I addressed your comments. Waiting for the extra reviewer

I can safely say @Mateoc will be ok with the constructor changes, but better if we get confirmation.

toddbaert reviewed (approved):
Looks great.

I'm not sure if you have any strong opinion on this, but though this does fix some behavior, it also adds significant new features. I wonder if `feat!` would be a better commit type?

I have removed "launch darkly client" from the title since our release flow automatically updates the readmes/release in such a way it's clear which component the commit applies to.

kinyoklion commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -34,73 +38,110 @@ export class LaunchDarklyClientProvider implements Provider {
+  async initialize(): Promise<void> {
+    /* for simplicity always initialize with anonymous by default*/
+
+    this._client = initialize(this.envKey, { anonymous: true }, this.ldOptions);
```
I don't find initializing with an anonymous context, unconditionally, to be a good idea. It generates less than optimal traffic (possibly an extra stream, and always an extra poll). Additionally it may cause substantial confusion with bootstrapping, as you ideally want that data to be associated with the correct context.

toddbaert replied:
Would your recommended solution be to make this a required part of the configuration then?

kinyoklion replied:
So, before, it allowed init with a context, and then needed to do a comparison. This was because the client existed before the provider.

Here I am confused though, I think this should just use the context that should be passed to `initialize`. Then the user would just need to set a global context before setting the provider? In which case it would just get initialized with that context?

sago2k8 replied:
Yeap I think I focus on my use case which is a error obviously, fixing it now, thanks @kinyoklion

kinyoklion commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -34,73 +38,110 @@ export class LaunchDarklyClientProvider implements Provider {
+     * use LD property streaming, since when enabled you are opting in for straming updates
+     * */
+    if (this.ldOptions?.streaming) {
+      this.setListeners();
```
This is probably fine, but the SDK itself would typically also start streaming if anyone listened for updates. So, this is kind of an indirect way of opting in.

Because of that I think we should add a note to the readme indicating that events will only be emitted if streaming is enabled.

toddbaert replied:
> This is probably fine, but the SDK itself would typically also start streaming if anyone listened for updates. So, this is kind of an indirect way of opting in.
> 
> Because of that I think we should add a note to the readme indicating that events will only be emitted if streaming is enabled.

In that case, we should probably add a note to the readme as you say, AND extend/override the jsdoc on the configuration object (I think you can add a link to the overridden jsdoc as well).

cc @sago2k8

sago2k8 replied:
I did added some more context in the README but nothing really in the jsdoc, do you happen to have an example on how to override the jsdoc properly?

sago2k8 commented:
Thanks for the review @kinyoklion, would you review it again ?

sago2k8 commented:
> Looks great.
> 
> I'm not sure if you have any strong opinion on this, but though this does fix some behavior, it also adds significant new features. I wonder if `feat!` would be a better commit type?
> 
> I have removed "launch darkly client" from the title since our release flow automatically updates the readmes/release in such a way it's clear which component the commit applies to.

Agree, set title to feat!

kinyoklion commented on libs/providers/launchdarkly-client/README.md:
```diff
@@ -11,34 +11,58 @@ $ npm install @openfeature/launchdarkly-client-provider
+/*
+ * You can set the initial context, this would be the context you pass to the LD initialize function
+ */
+const inialLDContext = {
```
`inialLDContext` -> `initialLDContext`.

kinyoklion commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -34,73 +38,113 @@ export class LaunchDarklyClientProvider implements Provider {
+  async initialize(): Promise<void> {
+    /* for simplicity always initialize with anonymous by default*/
+
+    this._client = initialize(this.envKey, this.initiaContext, this.ldOptions);
```
`initiaContext` -> `initialContext`.

kinyoklion commented on libs/providers/launchdarkly-client/src/lib/launchdarkly-client-provider.ts:
```diff
@@ -34,73 +38,113 @@ export class LaunchDarklyClientProvider implements Provider {
-    const ignoreAnonymous = newContext.anonymous && oldContext.anonymous && !newContext.key
-    if(!ignoreAnonymous && !isEqual(newContext, oldContext)) {
-       await this.client.identify(newContext);
+  async initialize(): Promise<void> {
```
I am potentially still missing something, but I think this takes a context, and my expectation is that context is used.


`  async initialize(context: EvaluationContext): Promise<void> {
	this._client = initialize(this.envKey, this.translateContext(evaluationContext), this.ldOptions);
`

sago2k8 replied:
You are completely right I was overlooking an important sdk behaviour, I think it's alright now, could you check it out @kinyoklion

toddbaert commented:
@sago2k8 @Mateoc related PR for ownership of this component: https://github.com/open-feature/js-sdk-contrib/pull/446/files#diff-0829329780a5428e2182662b54354b425c340a08b0b4468048abce4bd9d0312bR23-R26

toddbaert commented:
@sago2k8 I'm merging main into this to fix the CI issue.

kinyoklion reviewed (approved):
Minor comment, but I think things look good.

toddbaert commented:
I'll give @sago2k8 some time to address any nits or make final changes. Will merge tomorrow EOD unless I hear otherwise.

sago2k8 commented:
Hey @toddbaert I don't think I have received and invitation to an org yet.

toddbaert commented:
> Hey @toddbaert I don't think I have received and invitation to an org yet.

Very sorry @sago2k8 ! I added you and then removed you again when I alphabetized :facepalm: .

I've opened another PR to add you. When [this](https://github.com/open-feature/community/pull/228) is merged, you will get an invite.
